### PR TITLE
Implement Ledger Live bridge

### PR DESCRIPTION
**Pull request by darkwing**
Supercedes https://github.com/MetaMask/metamask-extension/pull/10171

This depends upon:
* https://github.com/MetaMask/eth-ledger-bridge-keyring/pull/55
* https://github.com/MetaMask/eth-ledger-bridge-keyring/pull/54

## Todo:

- [x] Add a setting to opt into this new functionality; if not, fall back to U2F
- [x] Implement fallback option vs. new functionality
- [x] Ensure design and verbiage integrity

### CODE REVIEW

File: ui/app/pages/create-account/connect-hardware/select-hardware.js
Comment (darkwing) [inline ui/app/pages/create-account/connect-hardware/select-hardware.js:164]: Needs l10n

File: ui/app/pages/create-account/connect-hardware/select-hardware.js
Comment (darkwing) [inline ui/app/pages/create-account/connect-hardware/select-hardware.js:209]: Needs L10n

File: ui/app/pages/settings/advanced-tab/advanced-tab.component.js
Comment (darkwing) [inline ui/app/pages/settings/advanced-tab/advanced-tab.component.js:396]: Needs L10n

File: ui/app/pages/settings/advanced-tab/advanced-tab.component.js
Comment (darkwing) [inline ui/app/pages/settings/advanced-tab/advanced-tab.component.js:398]: Needs L10n

File: app/scripts/metamask-controller.js
Comment (Gudahtt) [inline app/scripts/metamask-controller.js:2562]: Nit: We should follow this rule, rather than disable it

File: ui/app/pages/create-account/connect-hardware/account-list.js
Comment (Gudahtt) [inline ui/app/pages/create-account/connect-hardware/account-list.js:80]: Nit: We should follow this rule, rather than disable it

File: package.json
Comment (darkwing) [inline package.json:83]: This will obviously need to be changed once we publish changes on `metamask/eth-ledger-bridge-keyring`

File: ui/app/pages/settings/advanced-tab/advanced-tab.component.js
Comment (Gudahtt) [inline ui/app/pages/settings/advanced-tab/advanced-tab.component.js:408]: What is the connection between the Ledger Live bridge and WebUSB? I thought one of the main advantages of this was that it didn't depend on WebUSB, so it would work on Firefox.

Comment (darkwing) [reply ui/app/pages/settings/advanced-tab/advanced-tab.component.js]: Good shout -- removed!

File: app/scripts/metamask-controller.js
Comment (Gudahtt) [reply app/scripts/metamask-controller.js:2701]: Not sure why this was marked as resolved :thinking:

File: app/scripts/metamask-controller.js
Comment (Gudahtt) [inline app/scripts/metamask-controller.js:2688]: Is there any blocker to this being ab async function then? It seems like it should be.

File: app/scripts/metamask-controller.js
Comment (darkwing) [reply app/scripts/metamask-controller.js:2706]: Sorry, I likely marked it as resolved because we're using this pattern throughout the controller.  I can remove it but it seems like we need a larger effort to remove it completely.

Comment (Gudahtt) [reply app/scripts/metamask-controller.js:2706]: Oh I see - I hadn't realized. I mention because I've seen this used before because of an apparent conflict between this rule and the consistent return rule. But it has an easier workaround in cases like this where you want to return nothing: `return undefined`

File: app/scripts/metamask-controller.js
Comment (darkwing) [reply app/scripts/metamask-controller.js:2693]: Good shout!

File: ui/app/store/actions.js
Comment (brad-decker) [inline ui/app/store/actions.js:2826]: We really need to create a distinction by filename or something when dealing with actions like this that do not require redux or a thunk. In many PRs from many authors (myself included) this pattern has emerged of dispatching an action that only must be dispatched because it is constructed as a thunk like this.

Perhaps we need like a `background-actions.js` file for interacting with the promisifiedBackground? cc @Gudahtt

formal [REQ]: This doesn't need to be a traditional redux action. Return the call to `promisifed.setLedgerLivePreference` directly without the intermediary function. This will also mean that this action doesn't need to be included in `mapDispatchToProps` and can be imported and used directly.

**Review by brad-decker — CHANGES_REQUESTED** [review]

**Comment by brad-decker** [comment]
I just spent some time using this and it works like a dream on my Mac in Chrome. 😍

**Comment by brad-decker** [comment]
we should probably throw the "DO NOT MERGE" label on this until the reference to your fork for eth-ledger-bridge-keyring is updated to the new version of that library.

**Comment by darkwing** [comment]
Hello @rachelcope .  Screenshots as promised.  I did have to slightly change the text because we need to accommodate this feature being (1) turned on by the user and (2) turned off, thus the old way.

## "Connect Hardware" with nothing selected

LedgerLiveNoneSelected


## Trezor selected

TrezorSelected


## Ledger selected, new Ledger Live turned on

LedgerSelectedWithLedgerLiveSettingOn

## Ledger with Ledger Live off

LedgerPreferenceOff

**Review by rachelcope — DISMISSED** [review]
Looks great to me!

File: app/scripts/metamask-controller.js
Comment (Gudahtt) [inline app/scripts/metamask-controller.js:2691]: If this call fails, the UI would show no indication of failure. What's worse, the toggle state would not reflect the keyring state.

Ideally we'd `await` this call, so that we could forward the error to the UI. Ideally we'd try to keep it in-sync with the preference as well (e.g. maybe add a `catch` block where the previous preference value is restored? Or maybe set the preference only of this call succeeds? Something like that).

File: ui/app/pages/create-account/connect-hardware/account-list.js
Comment (Gudahtt) [inline ui/app/pages/create-account/connect-hardware/account-list.js:76]: This needs to be localized

Comment (darkwing) [reply ui/app/pages/create-account/connect-hardware/account-list.js:76]: D'oh, I added the localization to the `en` file but didn't use it.  Ugh.

File: app/scripts/metamask-controller.js
Comment (darkwing) [reply app/scripts/metamask-controller.js]: I've added the fallback.  Unfortunately, however, we should not `await` this because it's also called with `submitPassword`, and would delay the entire first metamask login/load until GitHub was happy.

File: ui/app/pages/create-account/connect-hardware/select-hardware.js
Comment (Gudahtt) [inline ui/app/pages/create-account/connect-hardware/select-hardware.js:143]: Nit: We should set the `key` prop here instead of disabling this rule. Not that the rule matters in this case, but better to avoid disabling rules in general if it's just as easy not to.

The same goes for the other disable comments in this file as well.

File: ui/app/pages/create-account/connect-hardware/select-hardware.js
Comment (Gudahtt) [inline ui/app/pages/create-account/connect-hardware/select-hardware.js:150]: This requires localization

File: app/_locales/en/messages.json
Comment (Gudahtt) [inline app/_locales/en/messages.json:1691]: Nit: In general we should include a `description` key for _all_ of these messages (we had agreed to commit to this as a team, but the linter hasn't been updated yet to reflect that), but we _especially_ should have descriptions whenever a substitution is used, to explain what will be substituted for `$1`.

The same goes for the other messages as well - I noticed a few others use substitutions.

File: ui/app/pages/create-account/connect-hardware/select-hardware.js
Comment (Gudahtt) [inline ui/app/pages/create-account/connect-hardware/select-hardware.js:177]: This ref was only set for the purposes of auto-scrolling, which was removed (see the removed `scrollToTutorial` function).  This can be removed as well. The same goes for the second copy of this below as well.

File: ui/app/pages/create-account/connect-hardware/select-hardware.js
Comment (Gudahtt) [inline ui/app/pages/create-account/connect-hardware/select-hardware.js:175]: Nit: It'd be nice to avoid duplicating this part between both of the tutorial functions.

Maybe there's a better way to organize this :thinking: This whole "steps" thing doesn't seem very idiomatic. Rather than populating this array and using it to generate the steps, it might be better to return the rendered markup directly, or with the help of additional render helper methods like `renderStep`, whatever makes sense.

File: ui/app/pages/create-account/connect-hardware/index.scss
Comment (Gudahtt) [inline ui/app/pages/create-account/connect-hardware/index.scss:122]: Is this necessary? :thinking: This is only being used on `<a>` tags as far as I can tell, which should already have a pointer cursor.

File: ui/app/pages/create-account/connect-hardware/index.scss
Comment (Gudahtt) [inline ui/app/pages/create-account/connect-hardware/index.scss:121]: Ideally we'd stick to using colours that are defined in our design system, in `ui/app/css/design-system/colors.scss`.

This exact colour doesn't seem to be there though :thinking:. Was this taken from the designs? Perhaps we can use one of the other types of blue instead.

File: ui/app/pages/create-account/connect-hardware/index.js
Comment (Gudahtt) [inline ui/app/pages/create-account/connect-hardware/index.js:326]: This appears to be unused

File: app/_locales/en/messages.json
Comment (Gudahtt) [inline app/_locales/en/messages.json:1545]: The addition of a substitution here will break all of our `selectAnAccountHelp` translations. We should remove them all.

File: ui/app/pages/create-account/connect-hardware/account-list.js
Comment (Gudahtt) [inline ui/app/pages/create-account/connect-hardware/account-list.js:71]: We should not be using `span` for clickable things if at all possible to avoid. We can use something like an `<a>` element instead, so that it has better accessibility properties built-in.

File: ui/app/pages/create-account/connect-hardware/index.scss
Comment (Gudahtt) [reply ui/app/pages/create-account/connect-hardware/index.scss:122]: Ah, I found an exception. It would still be better to use clickable elements and remove this rule though.

File: ui/app/pages/create-account/connect-hardware/index.scss
Comment (Gudahtt) [inline ui/app/pages/create-account/connect-hardware/index.scss:117]: I would really appreciate if we tried to stick to BEM-compatible classnames, which are of the format "[block]__element--modifier". This classname how has two `__` delimiters, which isn't valid BEM. Instead I'd suggest `-link` here instead of `__link`.

It's a pervasive pre-existing pattern I know,  but I've been trying not to let it spread further.

File: ui/app/pages/create-account/connect-hardware/account-list.js
Comment (Gudahtt) [inline ui/app/pages/create-account/connect-hardware/account-list.js:222]: This doesn't seem like the appropriate classname :eyes: 

I see it's used for the second button as well

File: ui/app/pages/create-account/connect-hardware/account-list.js
Comment (Gudahtt) [inline ui/app/pages/create-account/connect-hardware/account-list.js:124]: I'm curious what motivated this change. The most recent Ledger Live designs I saw still used the first and last 4 characters of the address, rather than 5.

File: app/scripts/controllers/preferences.js
Comment (darkwing) [reply app/scripts/controllers/preferences.js:677]: I wanted to follow the existing pattern, as seen by `setIpfsGateway`, `completeOnboarding`, etc.

Comment (Gudahtt) [reply app/scripts/controllers/preferences.js:677]: Ah I see. I think those were mistakes as well. Returning `Promise.resolve` from an otherwise synchronous function is bad.

If we want it to be async for compatibility reasons (e.g. maybe we want to leave the door open to do async things in here later or something), we should instead make the whole function `async` so that we have consistent error handling behaviour as well, rather than now where exceptions would get thrown synchronously rather than returned with a `Promise.reject`, which is _super_ unexpected/unintuitive for the caller.

But I don't see any case for this being async at all in this specific case. It seems inappropriate in the other examples as well.

File: ui/app/components/ui/toggle-button/toggle-button.component.js
Comment (darkwing) [reply ui/app/components/ui/toggle-button/toggle-button.component.js:57]: Good shout, left from a previous strategy!

File: ui/app/pages/create-account/create-account.component.js
Comment (darkwing) [reply ui/app/pages/create-account/create-account.component.js:45]: ++

File: ui/app/pages/create-account/connect-hardware/select-hardware.js
Comment (darkwing) [reply ui/app/pages/create-account/connect-hardware/select-hardware.js:143]: ++

File: ui/app/pages/create-account/connect-hardware/select-hardware.js
Comment (darkwing) [reply ui/app/pages/create-account/connect-hardware/select-hardware.js:150]: You're probably correct that "App" isn't universal, updated.

File: app/_locales/en/messages.json
Comment (darkwing) [reply app/_locales/en/messages.json:1691]: ++

File: ui/app/pages/create-account/connect-hardware/select-hardware.js
Comment (darkwing) [reply ui/app/pages/create-account/connect-hardware/select-hardware.js:177]: ++

File: ui/app/pages/create-account/connect-hardware/index.scss
Comment (darkwing) [reply ui/app/pages/create-account/connect-hardware/index.scss:121]: Good shout!

File: ui/app/pages/create-account/connect-hardware/index.js
Comment (darkwing) [reply ui/app/pages/create-account/connect-hardware/index.js:326]: Good catch!

File: app/_locales/en/messages.json
Comment (darkwing) [reply app/_locales/en/messages.json:1545]: I'll separate that sentence so we don't lose all our other translations; that's a valuable localized sentence.

File: ui/app/pages/create-account/connect-hardware/account-list.js
Comment (darkwing) [reply ui/app/pages/create-account/connect-hardware/account-list.js:71]: Good point, changed to button.

File: ui/app/pages/create-account/connect-hardware/index.scss
Comment (darkwing) [reply ui/app/pages/create-account/connect-hardware/index.scss:117]: Updated!

File: ui/app/pages/create-account/connect-hardware/account-list.js
Comment (darkwing) [reply ui/app/pages/create-account/connect-hardware/account-list.js:124]: Relic of the original PR, restored to 4.

File: app/_locales/en/messages.json
Comment (Gudahtt) [reply app/_locales/en/messages.json]: Huh, interesting solution! That works

File: ui/app/pages/create-account/connect-hardware/select-hardware.js
Comment (Gudahtt) [inline ui/app/pages/create-account/connect-hardware/select-hardware.js:144]: This class has been renamed to `hw-connect__msg-link`. I saw three uses of the old classname in this file

Comment (darkwing) [reply ui/app/pages/create-account/connect-hardware/select-hardware.js]: Updated, sorry!  /me grabs more coffee.

File: ui/app/pages/create-account/connect-hardware/index.scss
Comment (Gudahtt) [reply ui/app/pages/create-account/connect-hardware/index.scss:122]: I think this can safely be removed now? This is no longer used on non-pointer elements
```suggestion
```

File: ui/app/pages/create-account/connect-hardware/index.scss
Comment (Gudahtt) [inline ui/app/pages/create-account/connect-hardware/index.scss:134]: Huh - I think this is an unused classname :thinking: `hw-connect-link` isn't used anywhere.

**Review by Gudahtt — COMMENTED** [review]
Code looks good to me! Left a few more comments, but nothing super important left to address.

Holding off on approval for now until I've had the chance to test this out manually - I'll do that now.

File: ui/app/pages/create-account/connect-hardware/index.scss
Comment (darkwing) [reply ui/app/pages/create-account/connect-hardware/index.scss:134]: ++

File: ui/app/pages/create-account/connect-hardware/index.scss
Comment (darkwing) [reply ui/app/pages/create-account/connect-hardware/index.scss]: ++

File: app/scripts/metamask-controller.js
Comment (Gudahtt) [inline app/scripts/metamask-controller.js:2697]: W should probably re-throw the error here, so that the caller of `setLedgerLivePreference` can see that something failed.

File: ui/app/pages/create-account/connect-hardware/index.js
Comment (Gudahtt) [reply ui/app/pages/create-account/connect-hardware/index.js:147]: Huh. So I tried using the Ledger Live integration, and predictably it didn't work (the `ledger-live://` protocol isn't recognized on Linux), but I didn't see a timeout error either. It kept trying to connect forever.

Comment (darkwing) [reply ui/app/pages/create-account/connect-hardware/index.js:147]: With U2F, the same timeout should be used -- nothing changes here.  The timeout for ledger live is 3 minutes -- support asked for a larger amount of time: (https://github.com/MetaMask/eth-ledger-bridge-keyring/pull/54/files#diff-c0bdedb407c27ec7b9cc5c2e4a8d1d9a041eafa4a92c66b43f70c6e1c234af45R13)

With regard too `Internal JSON-RPC error`, that's not desirable; error messages shouldn't have changed.  I tested on our current released production version of MM (9.4.0) and I see the appropriate error when the Ledger is connected to my computer but not unlocked.  I checked out current `develop`, without my `eth-ledger-bridge-keyring` version update and the rest of the Ledger Live stuff, and also see "Internal JSON-RPC error." when the Ledger isn't unlocked.  

Could you confirm the same?  Any ideas on how we'd get here?

Comment (darkwing) [reply ui/app/pages/create-account/connect-hardware/index.js:147]: My update to `gh-pages` is hitting the correct error:

WrongAppErrorReturned

Comment (Gudahtt) [reply ui/app/pages/create-account/connect-hardware/index.js:147]: Interesting. I'll give this another try!

Comment (Gudahtt) [reply ui/app/pages/create-account/connect-hardware/index.js:147]: I tested on `develop` and saw the same "forever loading" when trying to connect via U2F. So that appears to be pre-existing at least. Similarly, when trying to switch pages with the Ethereum app closed, I see the "Internal JSON-RPC error" as well.

**Review by Gudahtt — DISMISSED** [review]
The code changes and the new UI get my approval! It looks like the error handling doesn't quite work (I'm still seeing `Internal JSON-RPC error` when the connection times out), but it's equally broken on `develop`, so that can be addressed separately.

I wasn't able to test the actual Ledger Live connection because it still doesn't seem to work on Linux, so I'll leave that up to someone else to test.

**Comment by brad-decker** [comment]
After a bunch of grief (which I believe to be a combination of yarn and mac os x's fault) I have now tested the following scenarios:
1. Switching back and forth between ledger live and U2F and performing a transaction each time
2. performing ledger live transaction while ledger live is closed using the popup UI (it opens, prompts to instantiate bridge, and then transaction appears on device) the extension popup UI closes during this but the transaction is handled properly.
3. Performing ledger live transaction in popup UI while bridge is connected -- works like a charm doesn't appear any different than current behavior
4. Performing ledger live transaction in popup UI while ledger live is open but bridge not established -- popup UI closes but upon opening bridge the transaction appears on device and all is good. 
5. All those variations on full screen UI, which works similarly but never closes due to losing focus.


more on my particular issues:
1. I've been working in eth-ledger-bridge-keyring, eth-keyring-controller, etc. 
2. I've been using yarn link judiciously
3. I've tested this branch at least 3 other times, so my state is not clean like a user who will be experiencing this for the first time in its production quality release

I kept having unexplainable issues related to eth-ledger-live-keyring. I even sent my index.js file (from node_modules in extension) to @darkwing and he was able to use my file flawlessly. 

After restarting my computer its all good and green and I have tested it thoroughly. 🤷🏻 
I've sent about twenty transactions now in various scenarios without a single hiccup.

**Review by brad-decker — DISMISSED** [review]
QA looks good to me can't wait to use this!

**Comment by darkwing** [comment]
This last commit is a reordering of commits to not conflict with @NiranjanaBinoy's `from-import-ui.spec.js`

**Review by brad-decker — APPROVED** [review]

**Pull request merged**
